File: internal/bff-service/server/http/handler/v1/skill_square.go
package v1

import (
	"fmt"
	"net/http"
	"net/url"

	"github.com/UnicomAI/wanwu/internal/bff-service/model/request"
	"github.com/UnicomAI/wanwu/internal/bff-service/service"
	gin_util "github.com/UnicomAI/wanwu/pkg/gin-util"
	"github.com/gin-gonic/gin"
)

// GetSquareBuiltinSkillList
//
//	@Tags			exploration.skill
//	@Summary		获取广场内置skill列表
//	@Description	获取探索广场中的内置skill列表
//	@Security		JWT
//	@Accept			json
//	@Produce		json
//	@Param			name	query		string	false	"skill名称"
//	@Success		200		{object}	response.Response{data=response.ListResult{list=[]response.BuiltinSkillInfo}}
//	@Router			/square/skill/builtin/list [get]
func GetSquareBuiltinSkillList(ctx *gin.Context) {
	resp, err := service.GetSquareBuiltinSkillList(ctx, getUserID(ctx), getOrgID(ctx), ctx.Query("name"))
	gin_util.Response(ctx, resp, err)
}

// GetSquareBuiltinSkillDetail
//
//	@Tags			exploration.skill
//	@Summary		获取广场内置skill详情
//	@Description	获取探索广场中的内置skill详情
//	@Security		JWT
//	@Accept			json
//	@Produce		json
//	@Param			skillId	query		string	true	"skill ID"
//	@Success		200		{object}	response.Response{data=response.BuiltinSkillDetail}
//	@Router			/square/skill/builtin/detail [get]
func GetSquareBuiltinSkillDetail(ctx *gin.Context) {
	resp, err := service.GetSquareBuiltinSkillDetail(ctx, ctx.Query("skillId"))
	gin_util.Response(ctx, resp, err)
}

// GetSquareShareSkillList
//
//	@Tags			exploration.skill
//	@Summary		获取广场共享skill列表
//	@Description	获取探索广场中的共享skill列表
//	@Security		JWT
//	@Accept			json
//	@Produce		json
//	@Param			name	query		string	false	"skill名称"
//	@Success		200		{object}	response.Response{data=response.ListResult{list=[]response.SharedSkillInfo}}
//	@Router			/square/skill/share/list [get]
func GetSquareShareSkillList(ctx *gin.Context) {
	resp, err := service.GetSquareShareSkillList(ctx, getUserID(ctx), getOrgID(ctx), ctx.Query("name"))
	gin_util.Response(ctx, resp, err)
}

// ShareSquareSkill
//
//	@Tags			exploration.skill
//	@Summary		添加共享skill到资源库
//	@Description	将共享skill添加到我的资源库
//	@Security		JWT
//	@Accept			json
//	@Produce		json
//	@Param			skillId	body		request.ShareSquareSkillReq	true	"skill ID"
//	@Success		200		{object}	response.Response
//	@Router			/square/skill/share [post]
func ShareSquareSkill(ctx *gin.Context) {
	var req request.ShareSquareSkillReq
	if !gin_util.Bind(ctx, &req) {
		return
	}
	err := service.ShareSquareSkill(ctx, getUserID(ctx), getOrgID(ctx), req.SkillId)
	gin_util.Response(ctx, nil, err)
}

// GetSquareShareSkillDetail
//
//	@Tags			exploration.skill
//	@Summary		获取共享skill详情
//	@Description	获取共享skill详情
//	@Security		JWT
//	@Accept			json
//	@Produce		json
//	@Param			skillId	query		string	true	"skill ID"
//	@Success		200		{object}	response.Response{data=response.SharedSkillDetail}
//	@Router			/square/skill/share/detail [get]
func GetSquareShareSkillDetail(ctx *gin.Context) {
	skillId := ctx.Query("skillId")
	resp, err := service.GetSquareShareSkillDetail(ctx, getUserID(ctx), getOrgID(ctx), skillId)
	gin_util.Response(ctx, resp, err)
}

// DownloadSquareShareSkill
//
//	@Tags			exploration.skill
//	@Summary		下载共享skill
//	@Description	下载共享skill ZIP包
//	@Security		JWT
//	@Accept			json
//	@Produce		application/octet-stream
//	@Param			skillId	query		string	true	"skill ID"
//	@Success		200		{object}	response.Response
//	@Router			/square/skill/share/download [get]
func DownloadSquareShareSkill(ctx *gin.Context) {
	skillId := ctx.Query("skillId")
	fileName := fmt.Sprintf("%s.zip", skillId)
	resp, err := service.DownloadSquareShareSkill(ctx, skillId)
	if err != nil {
		gin_util.Response(ctx, nil, err)
		return
	}
	ctx.Header("Content-Disposition", "attachment; filename*=utf-8''"+url.QueryEscape(fileName))
	ctx.Header("Content-Type", "application/octet-stream")
	ctx.Header("Access-Control-Expose-Headers", "Content-Disposition")
	ctx.Data(http.StatusOK, "application/octet-stream", resp)
}

// GetSquareShareSkillVersionList
//
//	@Tags			exploration.skill
//	@Summary		获取共享skill版本列表
//	@Description	获取共享skill的版本历史列表
//	@Security		JWT
//	@Accept			json
//	@Produce		json
//	@Param			skillId	query		string	true	"skill ID"
//	@Success		200		{object}	response.Response{data=response.ListResult{list=[]response.SkillVersionInfo}}
//	@Router			/square/skill/share/version/list [get]
func GetSquareShareSkillVersionList(ctx *gin.Context) {
	resp, err := service.GetSquareShareSkillVersionList(ctx, ctx.Query("skillId"))
	gin_util.Response(ctx, resp, err)
}

// GetSquareCreatedSkillList
//
//	@Tags			exploration.skill
//	@Summary		获取我发布skill列表
//	@Description	获取我发布的skill列表
//	@Security		JWT
//	@Accept			json
//	@Produce		json
//	@Param			name	query		string	false	"skill名称"
//	@Success		200		{object}	response.Response{data=response.ListResult{list=[]response.PublishedSkillInfo}}
//	@Router			/square/skill/created/list [get]
func GetSquareCreatedSkillList(ctx *gin.Context) {
	resp, err := service.GetSquareCreatedSkillList(ctx, getUserID(ctx), getOrgID(ctx), ctx.Query("name"))
	gin_util.Response(ctx, resp, err)
}

// GetSquareCreatedSkillDetail
//
//	@Tags			exploration.skill
//	@Summary		获取我发布skill详情
//	@Description	获取我发布的skill详情
//	@Security		JWT
//	@Accept			json
//	@Produce		json
//	@Param			customSkillId	query		string	true	"custom skill ID"
//	@Success		200				{object}	response.Response{data=response.PublishedSkillDetail}
//	@Router			/square/skill/created/detail [get]
func GetSquareCreatedSkillDetail(ctx *gin.Context) {
	resp, err := service.GetSquareCreatedSkillDetail(ctx, getUserID(ctx), getOrgID(ctx), ctx.Query("customSkillId"))
	gin_util.Response(ctx, resp, err)
}

// GetSquareCreatedSkillVersionList
//
//	@Tags			exploration.skill
//	@Summary		获取我发布skill版本列表
//	@Description	获取我发布的skill版本历史列表
//	@Security		JWT
//	@Accept			json
//	@Produce		json
//	@Param			customSkillId	query		string	true	"custom skill ID"
//	@Success		200				{object}	response.Response{data=response.ListResult{list=[]response.SkillVersionInfo}}
//	@Router			/square/skill/created/version/list [get]
func GetSquareCreatedSkillVersionList(ctx *gin.Context) {
	resp, err := service.GetSquareCreatedSkillVersionList(ctx, getUserID(ctx), getOrgID(ctx), ctx.Query("customSkillId"))
	gin_util.Response(ctx, resp, err)
}

File: web/src/api/skillSquare.js
// @description: skill广场相关接口
import request from '@/utils/request';
import { USER_API } from '@/utils/requestConstants';

// 获取skill广场列表-内置
export const getBuiltinSquareSkillList = data => {
  return request({
    url: `${USER_API}/square/skill/builtin/list`,
    method: 'get',
    params: data,
  });
};

// 发送skill广场到资源库
export const sendSquareSkillToResource = data => {
  return request({
    url: `${USER_API}/square/skill/share`,
    method: 'post',
    data,
  });
};
// 获取skill广场内置详情
export const getSquareSkillDetail = data => {
  return request({
    url: `${USER_API}/square/skill/builtin/detail`,
    method: 'get',
    params: data,
  });
};

// 下载skill广场skill
export const downloadSquareSkill = params => {
  return request({
    url: `${USER_API}/square/skill/download`,
    method: 'get',
    params,
    responseType: 'blob',
  });
};

// 获取广场共享skill列表
export const getSharedSquareSkillList = params => {
  return request({
    url: `${USER_API}/square/skill/share/list`,
    method: 'get',
    params,
  });
};

// 添加共享skill到资源库
export const addSharedSkillToResource = data => {
  return request({
    url: `${USER_API}/square/skill/share`,
    method: 'post',
    data,
  });
};

// 下载共享skill
export const downloadSharedSquareSkill = params => {
  return request({
    url: `${USER_API}/square/skill/share/download`,
    method: 'get',
    params,
    responseType: 'blob',
  });
};

// 获取共享skill详情
export const getSharedSquareSkillDetail = params => {
  return request({
    url: `${USER_API}/square/skill/share/detail`,
    method: 'get',
    params,
  });
};

// 获取我发布的 Skill 列表
export const getCreatedSquareSkillList = params => {
  return request({
    url: `${USER_API}/square/skill/created/list`,
    method: 'get',
    params,
  });
};

// 获取我发布的 Skill 详情
export const getCreatedSquareSkillDetail = params => {
  return request({
    url: `${USER_API}/square/skill/created/detail`,
    method: 'get',
    params,
  });
};

// 获取我发布的 Skill 版本列表
export const getCreatedSkillVersionList = params => {
  return request({
    url: `${USER_API}/square/skill/created/version/list`,
    method: 'get',
    params,
  });
};

// 获取共享 Skill 版本列表
export const getSharedSkillVersionList = params => {
  return request({
    url: `${USER_API}/square/skill/share/version/list`,
    method: 'get',
    params,
  });
};
